### PR TITLE
[SPARK-25274][PYTHON][SQL] In toPandas with Arrow send un-ordered record batches to improve performance

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

When executing `toPandas` with Arrow enabled, partitions that arrive in the JVM out-of-order must be buffered before they can be send to Python. This causes an excess of memory to be used in the driver JVM and increases the time it takes to complete because data must sit in the JVM waiting for preceding partitions to come in.

This change sends un-ordered partitions to Python as soon as they arrive in the JVM, followed by a list of partition indices so that Python can assemble the data in the correct order. This way, data is not buffered at the JVM and there is no waiting on particular partitions so performance will be increased.

Followup to #21546 

## How was this patch tested?

Added new test with a large number of batches per partition, and test that forces a small delay in the first partition. These test that partitions are collected out-of-order and then are are put in the correct order in Python.

## Performance Tests - toPandas

Tests run on a 4 node standalone cluster with 32 cores total, 14.04.1-Ubuntu and OpenJDK 8
measured wall clock time to execute `toPandas()` and took the average best time of 5 runs/5 loops each.

Test code
```python
df = spark.range(1 << 25, numPartitions=32).toDF("id").withColumn("x1", rand()).withColumn("x2", rand()).withColumn("x3", rand()).withColumn("x4", rand())
for i in range(5):
	start = time.time()
	_ = df.toPandas()
	elapsed = time.time() - start
```

Spark config
```
spark.driver.memory 5g
spark.executor.memory 5g
spark.driver.maxResultSize 2g
spark.sql.execution.arrow.enabled true
```

Current Master w/ Arrow stream | This PR
---------------------|------------
5.16207 | 4.342533
5.133671 | 4.399408
5.147513 | 4.468471
5.105243 | 4.36524
5.018685 | 4.373791

Avg Master | Avg This PR
------------------|--------------
5.1134364 | 4.3898886

Speedup of **1.164821449**

### CODE REVIEW

File: python/pyspark/serializers.py [hunk 1]
```diff
@@ -187,9 +187,15 @@ def loads(self, obj):
 
 class ArrowStreamSerializer(Serializer):
     """
-    Serializes Arrow record batches as a stream.
+    Serializes Arrow record batches as a stream. Optionally load the ordering of the batches as a
```
Comment (viirya) [inline python/pyspark/serializers.py]: This is optional. Do we have other usage of this `ArrowStreamSerializer` without the ordering?

Comment (BryanCutler) [reply python/pyspark/serializers.py]: Yeah, it's also used in the `createDataFrame` path, but that does only use `dump_stream`. Still, it seemed best to make this an optional feature of the serializer.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -3279,34 +3280,33 @@ class Dataset[T] private[sql](
...
-              batchWriter.end()
+            arrowBatches.indices.foreach { i => batchOrder.append((index, i)) }
+          }
+          partitionCount += 1
+
+          // After last batch, end the stream and write batch order
+          if (partitionCount == numPartitions) {
+            batchWriter.end()
+            out.writeInt(batchOrder.length)
+            // Batch order indices are from 0 to N-1 batches, sorted by order they arrived
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: How about a slight change? `// Re-order batches according to these indices to build a table.`

**Comment by BryanCutler** [comment]
Thanks @viirya !  What are your thoughts @HyukjinKwon ? I consolidated the batch order serializer from before into the ArrowStreamSerializer to simplify a little.

**Comment by BryanCutler** [comment]
@holdenk I was wondering if you had any thoughts on this? Thanks!

**Comment by holdenk** [comment]
Sure, I'll take a look on Friday if it's not urgent

**Review by felixcheung — COMMENTED** [review]
generally, is this going to limit how much data to pass along because of the bit length of the index?

**Comment by BryanCutler** [comment]
> generally, is this going to limit how much data to pass along because of the bit length of the index?

So the index passed to python is the RecordBatch index, not an element index, and it would limit the number of batches to Int.MAX. I wouldn't expect that would be likely and you can always set the number of batches to 1 per partition, so that would be the limiting factor then. WDYT @felixcheung ?

**Comment by felixcheung** [comment]
got it. so the size of the each batch could grow.

File: python/pyspark/serializers.py [hunk 3]
```diff
@@ -208,8 +214,26 @@ def load_stream(self, stream):
...
             yield batch
 
+        if self.load_batch_order:
+            num = read_int(stream)
+            self.batch_order = []
+            for i in xrange(num):
+                index = read_int(stream)
+                self.batch_order.append(index)
+
+    def get_batch_order_and_reset(self):
```
Comment (holdenk) [inline python/pyspark/serializers.py]: Looking at `_load_from_socket` I think I understand why this was done as a separate function here, but what about if the serializer its self returned either a tuple or re-ordered the batches its self?

I'm just trying to get a better understanding, not saying those are better designs.

File: python/pyspark/serializers.py [hunk 4]
```diff
@@ -208,8 +214,26 @@ def load_stream(self, stream):
         for batch in reader:
             yield batch
 
+        if self.load_batch_order:
+            num = read_int(stream)
+            self.batch_order = []
```
Comment (holdenk) [inline python/pyspark/serializers.py]: If we're going to have get_batch_order_and_reset as a separate function, could we verify batch_order is None before we reset and throw here if it's not? Just thinking of future folks who might have to debug something here.

File: python/pyspark/sql/tests.py [hunk 5]
```diff
@@ -4434,6 +4434,12 @@ def test_timestamp_dst(self):
         self.assertPandasEqual(pdf, df_from_python.toPandas())
         self.assertPandasEqual(pdf, df_from_pandas.toPandas())
 
+    def test_toPandas_batch_order(self):
+        df = self.spark.range(64, numPartitions=8).toDF("a")
+        with self.sql_conf({"spark.sql.execution.arrow.maxRecordsPerBatch": 4}):
+            pdf, pdf_arrow = self._toPandas_arrow_toggle(df)
+            self.assertPandasEqual(pdf, pdf_arrow)
```
[the remark this answers is not in the document]
Comment (holdenk) [reply python/pyspark/sql/tests.py]: This looks pretty similar to the kind of test case we could verify with something like hypothesis. Integrating hypothesis is probably too much work, but we could at least explore num partitions space in a loop quickly here. Would that help do you think @felixcheung ?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (holdenk) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: How about something like `// Sort by the output global batch indexes partition index, partition batch index tuple`?
When I was first read this code path I got confused my self so I think we should spend a bit of time on the comment here.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -3279,34 +3280,33 @@ class Dataset[T] private[sql](
...
-            // Write stored partitions that come next in order
-            while (lastIndex < results.length && results(lastIndex) != null) {
-              batchWriter.writeBatches(results(lastIndex).iterator)
-              results(lastIndex) = null
-              lastIndex += 1
-            }
-            // After last batch, end the stream
-            if (lastIndex == results.length) {
-              batchWriter.end()
+            arrowBatches.indices.foreach { i => batchOrder.append((index, i)) }
```
Comment (holdenk) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Could we call `i` something more descriptive like partition_batch_num or similar?

**Review by holdenk — COMMENTED** [review]
Thanks for working on this, reducing memory overhead pressure in the driver is always welcome :) I got a bit confused reading the PR the first time through so I think it might make sense to look at how we can improve the readability a bit.

File: python/pyspark/sql/tests.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (felixcheung) [reply python/pyspark/sql/tests.py]: that sounds good

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by BryanCutler** [comment]
Apologies for the delay in circling back to this. I reorganized a little to simplify and expanded the comments to hopefully better describe the code.

A quick summary of the changes: I changed the ArrowStreamSerializer to not have any state  - that seemed to complicate things. So instead of saving the batch order indices, they are loaded on the last iteration of `load_stream`, and this was put in a special serializer `ArrowCollectSerializer` so that it is clear where it is used.  I also consolidated all the batch ordering calls within `_collectAsArrow` so it is easier to follow the whole process.

File: python/pyspark/sql/tests.py [hunk 7]
```diff
@@ -4923,6 +4923,28 @@ def test_timestamp_dst(self):
...
+
+        cases = [
+            (1024, 512, 2),  # Try large num partitions for good chance of not collecting in order
+            (512, 64, 2),    # Try medium num partitions to test out of order collection
+            (64, 8, 2),      # Try small number of partitions to test out of order collection
+            (64, 64, 1),     # Test single batch per partition
+            (64, 1, 64),     # Test single partition, single batch
+            (64, 1, 8),      # Test single partition, multiple batches
+            (30, 7, 2),      # Test different sized partitions
+        ]
```
Comment (BryanCutler) [inline python/pyspark/sql/tests.py]: @holdenk and @felixcheung , I didn't do a loop but chose some different levels of partition numbers to be a bit more sure that partitions won't end up in order. I also added some other cases of different partition/batch ratios. Let me know if you think we need more to be sure here.

Comment (holdenk) [reply python/pyspark/sql/tests.py]: I don't see how we're guaranteeing out-of-order from the JVM. Could we delay on one of the early partitions to guarantee out of order?

Comment (BryanCutler) [reply python/pyspark/sql/tests.py]: Yeah, it's not a guarantee, but with a large num of partitions, it's a pretty slim chance they will all be in order. I can also add a case with some delay. My only concern is how big to make a delay to be sure it's enough without adding wasted time to the tests. 

How about we keep the case with a large number of partitions and add a case with 100ms delay on the first partition?

Comment (BryanCutler) [reply python/pyspark/sql/tests.py]: @holdenk , I updated the tests, please take another look when you get a chance. Thanks!

**Comment by BryanCutler** [comment]
ping @HyukjinKwon and @viirya to maybe take another look at the recent changes to make this cleaner, if you are able to. Thanks!

File: python/pyspark/sql/tests.py [hunk 8]
```diff
@@ -4923,6 +4923,34 @@ def test_timestamp_dst(self):
         self.assertPandasEqual(pdf, df_from_python.toPandas())
         self.assertPandasEqual(pdf, df_from_pandas.toPandas())
 
+    def test_toPandas_batch_order(self):
+
+        def delay_first_part(partition_index, iterator):
+            if partition_index == 0:
+                time.sleep(0.1)
```
Comment (holdenk) [inline python/pyspark/sql/tests.py]: I like this :)

File: python/pyspark/sql/tests.py (hunk 7 above)
Comment (holdenk) [reply python/pyspark/sql/tests.py]: I like the new tests, I think 0.1 on one of partitions is enough.

**Review by holdenk — APPROVED** [review]
I like the change to the tests, thanks @BryanCutler

**Comment by viirya** [comment]
LGTM the current change looks clearer. Thanks @BryanCutler

**Comment by BryanCutler** [comment]
merged to master, thanks @holdenk @viirya and @felixcheung !

**Comment by HyukjinKwon** [comment]
Sorry @BryanCutler for my super super late input. LGTM to me as well :D.

**Comment by BryanCutler** [comment]
Thanks @HyukjinKwon !

**Pull request closed without merging**